users/views.py

```diff
@@ -302,19 +307,61 @@ def get_context_data(self, **kwargs):
         return context
 
 
-class DeleteUserView(LoginRequiredMixin, SuccessMessageMixin, DeleteView):
+class DeleteUserView(LoginRequiredMixin, SuccessMessageMixin, FormView):
     template_name = "users/delete.html"
+    success_url = reverse_lazy("profile-account")
+    form_class = DeleteAccountForm
+
+    def get_object(self):
+        return self.request.user
+
+    def form_valid(self, form):
+        success_url = self.get_success_url()
+        user = self.get_object()
+        user.delete_permanently_at = timezone.now() + datetime.timedelta(
+            days=settings.ACCOUNT_DELETION_GRACE_PERIOD_DAYS
+        )
+        user.save()
+        return HttpResponseRedirect(success_url)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context[
+            "ACCOUNT_DELETION_GRACE_PERIOD_DAYS"
+        ] = settings.ACCOUNT_DELETION_GRACE_PERIOD_DAYS
+        return context
+
+
+class CancelDeletionView(LoginRequiredMixin, SuccessMessageMixin, FormView):
+    form_class = forms.Form
+    success_url = reverse_lazy("profile-account")
+    template_name = "users/cancel_deletion.html"
+    success_message = "Your account is no longer scheduled for deletion."
+
+    def get_object(self):
+        return self.request.user
+
+    def form_valid(self, form):
+        user = self.get_object()
+        user.delete_permanently_at = None
+        user.save()
+        return super().form_valid(form)
+
+
+class DeleteImmediatelyView(LoginRequiredMixin, SuccessMessageMixin, FormView):
+    form_class = DeleteAccountForm
+    template_name = "users/delete_immediately.html"
     success_url = "/"
     success_message = "Your profile was successfully deleted."
-    form_class = DeleteAccountForm
 
     def get_object(self):
         return self.request.user
 
     def form_valid(self, form):
```

Comment:
This should follow the same pattern you used in `CancelDeletionView`. No need to call get_success_url or add the message, or directly return a redirect. Leverage `super`.

Reply:
nice!